processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -127,37 +133,57 @@ protected Integer addToFacts(
       ThreadLocal<InputRow> in
   ) throws IndexSizeExceededException
   {
-    Integer rowOffset;
-    synchronized (this) {
-      rowOffset = numEntries.get();
-      if(rowOffset >= maxRowCount && !facts.containsKey(key)) {
+    Integer priorIdex = facts.get(key);
+    final Boolean hadPriorValue = null != priorIdex;
+
+    Aggregator[] aggs = hadPriorValue ? aggList.get(priorIdex) : new Aggregator[metrics.length];
+
+    if (!hadPriorValue) {
+      for (int i = 0; i < metrics.length; i++) {
+        final AggregatorFactory agg = metrics[i];
+        aggs[i] = agg.factorize(
+            makeColumnSelectorFactory(agg, in, deserializeComplexMetrics)
+        );
+      }
+      // Backfill holes
+      Integer rowIdex = indexIncrement.getAndIncrement();
```

Comment:
That's actually not a typo. I don't know where I picked up the habit but I usually call them blahIdex instead of blahIndex... I have no idea why really. I'll change it to make it more generally readable.